src/query/functions/temporal/linear_regression.go

```diff
@@ -0,0 +1,155 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package temporal
+
+import (
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/ts"
+)
+
+const (
+	// PredictLinearType predicts the value of time series t seconds from now, based on the input series, using simple linear regression.
+	// PredictLinearType should only be used with gauges.
+	PredictLinearType = "predict_linear"
+
+	// DerivType calculates the per-second derivative of the time series, using simple linear regression.
+	// DerivType should only be used with gauges.
+	DerivType = "deriv"
+)
+
+type linearRegressionProcessor struct {
+	duration float64
+}
+
+func (l linearRegressionProcessor) Init(op baseOp, controller *transform.Controller, opts transform.Options) Processor {
+	return &linearRegressionNode{
+		op:         op,
+		controller: controller,
+		timeSpec:   opts.TimeSpec,
+		duration:   l.duration,
+	}
+}
+
+// NewLinearRegressionOp creates a new base temporal transform for linear regression functions
+func NewLinearRegressionOp(args []interface{}, optype string) (transform.Params, error) {
+	var (
+		duration float64
+		ok       bool
+	)
+
+	switch optype {
+	case PredictLinearType:
+		if len(args) != 2 {
+			return emptyOp, fmt.Errorf("invalid number of args for %s: %d", PredictLinearType, len(args))
+		}
+
+		duration, ok = args[1].(float64)
+		if !ok {
+			return emptyOp, fmt.Errorf("unable to cast to scalar argument: %v for %s", args[1], PredictLinearType)
+		}
+
+	case DerivType:
+	default:
+		return nil, fmt.Errorf("unknown linear regression type: %s", optype)
+	}
+
+	l := linearRegressionProcessor{
+		duration: duration,
+	}
+
+	return newBaseOp(args, optype, l)
+}
+
+type linearRegressionNode struct {
+	op         baseOp
+	controller *transform.Controller
+	timeSpec   transform.TimeSpec
+	duration   float64
+}
+
+func (l linearRegressionNode) Process(dps ts.Datapoints, alignedTime time.Time) float64 {
+	if dps.Len() < 2 {
+		return math.NaN()
+	}
+
+	var slope, intercept float64
+
+	if l.op.operatorType == PredictLinearType {
```

Comment:
This reads a little weird to me, since you're splitting processing amongst two functions and doing related calculations in both.

Instead, can you pull the code from `linearRegression()` into this `Process` function, and let it do the switching on its own?

 Alternatively, if you want to avoid dealing with operator type inside the regression function, you can pass a bool that signifies wether you're calculating regression or deriv into there, and have it just return a float, something like `linearRegression(isDeriv bool, dps ts.Datapoints, interceptTime time.Time) float64`

Reply:
I like separate linearRegression since it makes it easy to test. My above comment about having a separate func instead duration should fix this problem.